internal/credential/state.go

```diff
@@ -269,29 +281,22 @@ func AwsCredentialPersistedStateFromProto(secrets *structpb.Struct, attrs *Crede
 	return s, nil
 }
 
-// GetCredentialType returns the credential type based on the given AWS AccessKey.
-//
-// https://docs.aws.amazon.com/IAM/latest/UserGuide/security-creds.html#sec-access-keys-and-secret-access-keys
-// Access key IDs beginning with ASIA are temporary credentials access keys that you create using AWS STS operations.
-//
-// https://docs.aws.amazon.com/IAM/latest/UserGuide/security-creds.html#sec-access-keys-and-secret-access-keys
-// Access key IDs beginning with AKIA are long-term access keys for an IAM user or an AWS account root user.
-func GetCredentialType(accessKey string) CredentialType {
-	if strings.HasPrefix(accessKey, "ASIA") {
-		return Dynamic
-	}
-	if strings.HasPrefix(accessKey, "AKIA") {
-		return Static
+// GetCredentialType returns the credential type based on the given
+// AccessKey/RoleARN. See CredentialType definition for more information.
+func GetCredentialType(cc *awsutilv2.CredentialsConfig) CredentialType {
+	if cc == nil {
+		return Unknown
 	}
-	return Unknown
-}
 
-// HasDynamicCredentials returns true if the access key is a dynamic credential type.
-func HasDynamicCredentials(accessKey string) bool {
-	return strings.HasPrefix(accessKey, "ASIA")
-}
+	if len(cc.RoleARN) > 0 || strings.HasPrefix(cc.AccessKey, "ASIA") {
```

Comment:
Is there a potential code path where the cc.RoleARN is empty, but the AccessKey with the prefix `ASIA` is true? I don't see a use case where this length check would be required. I think that checking the prefix will cover any scenario.

I checked the awsutil codebase and RoleARN value is set by using the `withRoleARN` flag or its pulled from environment variables.

Reply:
That is a good point. I think that we should keep the roleARN length check in the conditional check.